spec/06-expressions.md

```diff
@@ -1607,6 +1609,8 @@ finding a substitution $\sigma$ of types $T_i$ for the type parameters
 $a_i$ such that
 
 - None of the inferred types $T_i$ is a [singleton type](03-types.html#singleton-types)
+  unless it is a singleton type corresponding to an object or a constant value
+  definition or the corresponding bound $U_i$ is a subtype of `scala.Singleton`.
```

Comment:
Thanks adding a mention of `<: Singleton`. I think I misunderstood the text on my previous comment here; anyway, in my current interpretation, it sounds like a singleton type should always be inferred if it denotes an object/constant value.

Where does the spec say that `List[Int]` is inferred for the example below, not `List[1]`?

```
scala> def f[T](x: T): List[T] = Nil
f: [T](x: T)List[T]

scala> object t { final val x = f(1) }
defined object t

scala> t.x: List[1]
         ^
       error: type mismatch;
        found   : List[Int]
        required: List[1]
```

Reply:
More or less exactly in the area you're commenting on. The right hand side of `val x` is `f(1)` and the type parameter `T` of `f` at that call site will be inferred as the widened type ("None of the inferred types Ti is a singleton type unless ..."). If `x` had been given an explicit `: 1` annotation that would have driven the inference of `T`.

Reply:
Thanks, I'm getting there :) Shouldn't this type check then:

```
scala> :pa
// Entering paste mode (ctrl-D to finish)

def f[T](x: T): List[T] = Nil
object o
object a { final val x = o; final val y = 1 }
object t { final val x = f(a.x); final val y = f(a.y) }

// Exiting paste mode, now interpreting.

scala> a.x : o.type
res0: o.type = o$@b46e103

scala> a.y: 1
res2: Int = 1

scala> t.x : List[o.type]
res4: List[o.type] = List()

scala> t.y : List[1]
         ^
       error: type mismatch;
        found   : List[Int]
        required: List[1]
```

Reply:
No, in `f(a.y)` the inferred type for `T` is widened to `Int` because `T` doesn't have a `<: Singleton` bound.

Reply:
OK, I get it, the compiler knows `implicitly[a.x.type =:= o.type]`, so widening to `o.type` doesn't discard information (or, `scala> o : a.x.type` type checks).